perf/light.cc

```diff
@@ -0,0 +1,653 @@
+	// own but we have nothing left to do here to make it closer to
+	// Light bench).
+	struct TupleHash {
+		std::size_t operator()(const Tuple &t) const
+		{
+			return t.hash;
+		}
+	};
+}
+
+// Simple generator of uniformly distributed random bytes.
+struct RandomBytesGenerator {
+	RandomBytesGenerator() :
+		gen(std::random_device()()), distr(0, UCHAR_MAX) { }
+	unsigned char get() { return distr(gen); }
+	std::mt19937 gen;
+	std::uniform_int_distribution<unsigned short> distr;
+};
+
+// Large chunk of continuous memory initialized with random bytes.
+struct TupleHolder {
+	TupleHolder(std::size_t tuple_count) : mem(nullptr), tuples()
+	{
+		std::size_t total_size = tuple_count * TUPLE_SIZE;
+		mem = static_cast<unsigned char *>(malloc(total_size));
+		for (std::size_t i = 0; i < total_size; ++i) {
+			mem[i] = TupleHolder::random_generator.get();
+		}
+		tuples.reserve(tuple_count);
+		for (std::size_t i = 0; i < tuple_count; ++i) {
+			uint32_t key = mem[i * TUPLE_SIZE];
+			tuples.push_back(Tuple(&mem[i * TUPLE_SIZE], TUPLE_SIZE, key));
+		}
+	}
+
+	void shuffle(std::size_t count)
+	{
+		// Instead of generating new values in most cases it's enough
+		// to shuffle old ones.
+		std::random_shuffle(tuples.begin(), tuples.begin() + count);
+	}
+
+	TupleHolder(const TupleHolder &) = delete;
+	TupleHolder(TupleHolder &&) = delete;
+
+	~TupleHolder()
+	{
+		tuples.clear();
+		free(mem);
+		mem = nullptr;
+	}
+
+	unsigned char *mem;
+	std::vector<Tuple> tuples;
+	static RandomBytesGenerator random_generator;
+};
+
+RandomBytesGenerator TupleHolder::random_generator{};
+
+////////////////////////////// Light Definitions ///////////////////////////////////////////////////////////////////////
+
+bool
+value_equals(Value_t v1, Value_t v2)
+{
+	assert(v2.size == v1.size);
+	return std::memcmp(v1.data, v2.data, v1.size) == 0;
+}
+
+bool
+key_equals(Value_t k1, Key_t k2)
+{
+	return k1.data[0] == k2;
+}
+
+#define LIGHT_NAME
+#define LIGHT_DATA_TYPE Value_t
+#define LIGHT_KEY_TYPE uint8_t
+#define LIGHT_CMP_ARG_TYPE int
+#define LIGHT_EQUAL(a, b, arg) value_equals(a, b)
+#define LIGHT_EQUAL_KEY(a, b, arg) key_equals(a, b)
+#include "salad/light.h"
+
+enum {
+	KB = 1 << 10,
+};
+
+static constexpr std::size_t light_extent_size = 16 * KB;
+static std::size_t extents_count = 0;
+
+inline void *
+light_malloc_extend(void *ctx)
+{
+	std::size_t *p_extents_count = (size_t *)ctx;
+	assert(p_extents_count == &extents_count);
+	++*p_extents_count;
+	return malloc(light_extent_size);
+}
+
+inline void
+light_free_extend(void *ctx, void *p)
+{
+	size_t *p_extents_count = (size_t *) ctx;
+	assert(p_extents_count == &extents_count);
+	--*p_extents_count;
+	free(p);
+}
+
+////////////////////////////// Light Fixture ///////////////////////////////////////////////////////////////////////////
+
+class LightFixture : public benchmark::Fixture {
+public:
+	void
+	SetUp(const benchmark::State &state)
+	{
+		ht = static_cast<struct light_core *>(malloc(sizeof(*ht)));
```

Comment:
Why can't you create this object on the stack?

Reply:
Agree.

Reply:
This is fixed.